### PR TITLE
Fix orphan nodes on 3D meshes for  BreakMeshByBlockGenerator

### LINKED ISSUE #15793: BreakMeshByBlock fix orphan nodes on boundary
## Bug Description
When one of the original interface nodes remains orphan (e.g is not woned anymore by any element belonging to the same processor) triggers the following assertion
Assertion `!node_touched_by_anyone[nodeid] || node_touched_by_me[nodeid]' failed.

This is only triggered in debug mode when enough CPUs are used and with 3D meshes.

## Steps to Reproduce
I prepared a test triggering this problem in the solution PR

### PR DESCRIPTION
**Pull request by arovinelli**
When one of the original interface nodes remains orphan (e.g is not woned anymore by any element belonging to the same processor) triggers the following assertion
Assertion `!node_touched_by_anyone[nodeid] || node_touched_by_me[nodeid]' failed.

This is only triggered in debug mode when enough CPUs are used and with 3D meshes.

I fixed this using `Partitioner::set_node_processor_ids`

thanks @roystgnr for the help.

Closes #15793

### CODE REVIEW

**Comment by arovinelli** [comment]
@lindsayad this is ready for review when the CI comes back opnline

**Comment by arovinelli** [comment]
> Please push a change to this to forcefully re-trigger testing (full testing is now available).
> 
> You can do the following:
> 
> ```
> git commit --amend
> ```
> 
> and save the amend message that comes up without changing it (by typing `:wq` if in vim). This will forcefully change the commit hash of the last commit, which will trigger a change. Then forcefully push:
> 
> ```
> git push --force
> ```

tried that but it doesn't trigger tests to run

**Comment by milljm** [comment]
I am looking into the failure...

**Comment by milljm** [comment]
Got it. We are constantly added rules to our security efforts... Thanks!

**Comment by lindsayad** [comment]
@arovinelli Do you expect your test to work in recover mode?

**Comment by arovinelli** [comment]
@lindsayad I'm sorry for the delay. I removed a couple of commits as they are related to a new feature rather than to simply fixing the bug. I'll add a new PR for that. 
To answer your question, I'm not expecting `BreakMeshByBlock` to work in recover mode, hopefully in the future, (still need to fix distributed first). Anyway, this should be ready

File: test/tests/meshgenerators/break_mesh_by_block_generator/tests [hunk 1]
```diff
@@ -70,4 +70,19 @@
...
+    cli_args = '--mesh-only'
+    exodiff = 'hanging_nodes_parallel_in.e'
+    requirement = 'MOOSE shall be able to create separate blocks in a mesh in parallel'
+    design = 'meshgenerators/BreakMeshByBlockGenerator.md'
+    issues = '#15793'
+    mesh_mode = 'REPLICATED'
+    min_parallel = 4
+    max_parallel = 5
+    recover = false
+    method = 'dbg'
```
Comment (loganharbour) [inline test/tests/meshgenerators/break_mesh_by_block_generator/tests:86]:
```suggestion
```
Go ahead and remove this. It is a good idea to run it in opt too even though your assertion doesn't get hit there.

File: test/tests/meshgenerators/break_mesh_by_block_generator/hanging_nodes_parallel.i [hunk 2]
```diff
@@ -0,0 +1,69 @@
+[GlobalParams]
+  displacements = "disp_x disp_y disp_z"
+[]
+
```
Comment (loganharbour) [inline test/tests/meshgenerators/break_mesh_by_block_generator/hanging_nodes_parallel.i:4]:
```suggestion
```

**Review by loganharbour — CHANGES_REQUESTED** [review]
I attempted to add a `GridPartitioner` to you problem to guarantee on all OSes regardless of the default partitioner that you hit your issue here. But apparently the `[Partitioner]` block doesn't work with the `MeshGenerator` system? Looking into this...

File: test/tests/meshgenerators/break_mesh_by_block_generator/hanging_nodes_parallel.i (hunk 2 above)
Comment (arovinelli) [reply test/tests/meshgenerators/break_mesh_by_block_generator/hanging_nodes_parallel.i:4]: if I remove this block it doesn't hit the assertion

Comment (arovinelli) [reply test/tests/meshgenerators/break_mesh_by_block_generator/hanging_nodes_parallel.i]: moved displacements inside the mesh_block

**Comment by arovinelli** [comment]
@loganharbour if this passes tests to me is ready to be merged as the fix is global (`Partitioner::set_node_processor_ids` is called to check for orphan nodes and, if necessary, fix their owning CPU). I'm not sure we can add a partitioner block this for better coverage. On Centos OS, the proposed change fixes the issue for both `devel` and `debug` mode.

**Comment by arovinelli** [comment]
@loganharbour pinging so this PR will stay alive

**Comment by arovinelli** [comment]
this is not stale @loganharbour

**Comment by loganharbour** [comment]
@arovinelli I’ve added this to my list for the beginning of next week, I appreciate the reminder.

@lindsayad are RMs compatible with mesh generators now, or do I still need to do something on that front?

**Comment by lindsayad** [comment]
Yes they are compatible.

**Comment by arovinelli** [comment]
Don’t

**Comment by loganharbour** [comment]
Good news @arovinelli - once #16777 gets in, we can force a grid partitioner on this to make the test consistent. I should be able got give that a try tomorrow.

**Comment by arovinelli** [comment]
> Good news @arovinelli - once #16777 gets in, we can force a grid partitioner on this to make the test consistent. I should be able to give that a try tomorrow.

@loganharbour Great, thanks for letting me know! I'm really looking forward for it!

**Comment by arovinelli** [comment]
@loganharbour I see #16777 was merged. Besides rebasing and resolving conflicts is there anything else I need to do?

**Comment by arovinelli** [comment]
@loganharbour I added a commit forcing the default partitioner in the tests file, so this should be ready to go. Also, why tests are not running?

**Comment by lindsayad** [comment]
I'm guessing this PR occurred during the CIVET shutdown. You may have to
open a new one.

**Comment by arovinelli** [comment]
@lindsayad and @loganharbour don't know what happened, but tests are running now

**Comment by lindsayad** [comment]
> Not related to this PR

It probably is related to this PR if it was a SQA test failure

**Comment by lindsayad** [comment]
The Documentation target will tell us

File: test/tests/meshgenerators/break_mesh_by_block_generator/hanging_nodes_parallel.i [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+  []
+  [./split]
+    type = BreakMeshByBlockGenerator
+    input = subdomain_6
+  []
+[]
+
+[Debug]
+  pid_aux = true
+[]
```
Comment (loganharbour) [inline test/tests/meshgenerators/break_mesh_by_block_generator/hanging_nodes_parallel.i:71]:
```suggestion
```

Should be able to remove this now

File: test/tests/meshgenerators/break_mesh_by_block_generator/tests [hunk 4]
```diff
@@ -149,4 +149,17 @@
...
+    type = 'Exodiff'
+    input = 'hanging_nodes_parallel.i'
+    exodiff = 'hanging_nodes_parallel_out.e'
+    requirement = 'The system shall properly assign node processor ids for paritioned meshes'
+    design = 'meshgenerators/BreakMeshByBlockGenerator.md'
+    issues = '#15793'
+    mesh_mode = 'REPLICATED'
+    min_parallel = 12
+    max_parallel = 12
+    recover = false
```
Comment (loganharbour) [inline test/tests/meshgenerators/break_mesh_by_block_generator/tests:163]: I believe that you can likely remove this now that we don't have --mesh-only, right?

**Review by loganharbour — CHANGES_REQUESTED** [review]

File: test/tests/meshgenerators/break_mesh_by_block_generator/tests (hunk 4 above)
Comment (arovinelli) [reply test/tests/meshgenerators/break_mesh_by_block_generator/tests:163]: I can't remove 
```    mesh_mode = 'REPLICATED'
    recover = false``` becasue break mesh by block works only in replicated mode and currenlty deos not allow recover
    
also, if I remove 
```
    min_parallel = 12
    max_parallel = 12
```
and run the tests in debug mode nad run in debug with -j12 or -p12 will no trigger the assertion

so all those for lines should stay.

Comment (loganharbour) [reply test/tests/meshgenerators/break_mesh_by_block_generator/tests:163]: This comment was meant only at the recover option (it displays the lines above in addition). Sounds good

File: test/tests/meshgenerators/break_mesh_by_block_generator/hanging_nodes_parallel.i (hunk 3 above)
Comment (arovinelli) [reply test/tests/meshgenerators/break_mesh_by_block_generator/hanging_nodes_parallel.i:71]: Yes, I removed it and cahnged the test to RunApp

**Review by loganharbour — APPROVED** [review]
Thanks for spending the time to get the tests right!

**Comment by arovinelli** [comment]
> Thanks for spending the time to get the tests right!

you are welcome :)

**Review by lindsayad — APPROVED** [review]

**Comment by arovinelli** [comment]
pinging @jiangwen84, the fix for oprhan nodes in BMBB was merged

**Pull request merged**